### PR TITLE
Cut GetTypeInfo from ConcurrentDictionary and use switch over if-else

### PR DESCRIPTION
**Pull request by JonHanna**
`GetTypeInfo` is no longer a necessary step to get `IsValueType` so remove the call to it (contributes to #16305).

Replace if-else ladder on `== typeof(…)` with switch on `GetTypeCode` so up to a dozen tests becomes at most 3.

This makes it no longer necessary to test for enum types and get underlying type, so don't.

Add a 64-bit type to the types covered in testing.

### CODE REVIEW

File: src/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentDictionary.cs [hunk 1]
```diff
@@ -93,31 +93,31 @@ private static bool IsValueWriteAtomic()
...
+                case TypeCode.Char:
+                case TypeCode.Int16:
+                case TypeCode.Int32:
+                case TypeCode.SByte:
+                case TypeCode.Single:
+                case TypeCode.UInt16:
+                case TypeCode.UInt32:
+                    return true;
+                case TypeCode.Int64:
+                case TypeCode.DateTime:
```
Comment (stephentoub) [inline src/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentDictionary.cs]: Why did you add DateTime?

Comment (JonHanna) [inline src/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentDictionary.cs]: As a struct with a single 64-bit field it should be atomically writeable on 64-bit and we've already got its typecode at this point, so there doesn't seem to be much point in leaving it out.

Comment (stephentoub) [inline src/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentDictionary.cs]: @jkotas, any thoughts on this?  On the one hand, given that the backing state for DateTime is a ulong which should be atomically read/written on 64-bit, this should be safe, and it's unlikely DateTime's layout would change.  On the other hand, it makes me nervous that this code would take a dependency on such an implementation detail, DateTime isn't explicitly called out as being atomically read/written, etc.

Comment (jkotas) [inline src/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentDictionary.cs]: I agree that it is not a good practice to take dependency on undocumented internal implementation details, but it is unlikely for DateTime implementation to change.

Is it common to use DateTime with ConcurrentDictionary? I do not think it is, thus I would not bother writing fragile code for it.

Accidently, somebody asked me pretty same question last week in the context of https://stackoverflow.com/questions/42258346/can-datetime-tear-in-a-64-bit-environment . `AreReadsAndWritesAreGuaranteedAtomic` method given as "answer" is wrong. It is not possible to infer whether the struct assignment is atomic by just looking at its size. It would need to inspect the struct fields using reflection.

Comment (stephentoub) [inline src/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentDictionary.cs]: @JonHanna, let's undo the DateTime part of this.  If it proves to be very common as a value in the future, we can re-evaluate.

Comment (JonHanna) [inline src/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentDictionary.cs]: Sure. I very much doubt it would be that common, and I added it on the basis that once we're switching on type code adding that case is cheap. If there's any doubt around it I'd rather be cautious.

Comment (JonHanna) [inline src/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentDictionary.cs]: Hmm. That SO answer is also incorrect in not checking for a `StructLayoutAttribute`.
